src/GraphQL/Internal/AST.hs

```diff
@@ -45,24 +52,71 @@ module GraphQL.Internal.AST
 
 import Protolude hiding (Type)
 
-import Data.Char (isDigit)
+import qualified Data.Aeson as Aeson
 import qualified Data.Attoparsec.Text as A
-import GraphQL.Internal.Tokens (tok)
+import Data.Char (isDigit)
+import Test.QuickCheck (Arbitrary(..), elements, listOf)
 
--- TODO: Our GraphQL.Value.Name has more guarantees than AST.Name (which is
--- just a Text alias). We should change that so AST.Name is guaranteed valid.
+import GraphQL.Internal.Tokens (tok)
 
 -- * Name
 
-type Name = Text
+-- | A name in GraphQL.
+--
+-- https://facebook.github.io/graphql/#sec-Names
+newtype Name = Name { getNameText :: Text } deriving (Eq, Ord, Show)
```

Comment:
Ah I see, thanks for explaining! Not exporting the constructor is a good point. I think I'd then prefer `getName` or `unName` (like `unwrap` in https://github.com/purescript/purescript-newtype/blob/master/src/Data/Newtype.purs)